dll/win32/shell32/shelldesktop/CChangeNotify.cpp

```diff
@@ -0,0 +1,581 @@
+{
+    HWND hwndNotif = DoGetNewDeliveryWorker();
+    if (!hwndNotif)
+        return;
+
+    DWORD pid;
+    GetWindowThreadProcessId(hwndNotif, &pid);
+
+    HANDLE hTicket = DoCreateDeliTicket(wEventId, uFlags, pidl1, pidl2, pid, dwTick);
+    if (hTicket)
+    {
+        TRACE("hTicket: %p, 0x%lx\n", hTicket, pid);
+        if ((uFlags & (SHCNF_FLUSH | SHCNF_FLUSHNOWAIT)) == SHCNF_FLUSH)
+            SendMessageW(hwndNotif, WM_NOTIF_DELIVERY, (WPARAM)hTicket, pid);
+        else
+            SendNotifyMessageW(hwndNotif, WM_NOTIF_DELIVERY, (WPARAM)hTicket, pid);
+    }
+}
+
+/*static*/ LRESULT CALLBACK
+CWorker::WindowProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    CWorker *pThis = (CWorker *)GetWindowLongPtrW(hwnd, 0);
+    if (pThis)
+    {
+        LRESULT lResult = 0;
+        pThis->ProcessWindowMessage(hwnd, uMsg, wParam, lParam, lResult, 0);
+        return lResult;
+    }
+    return 0;
+}
+
+BOOL CWorker::CreateWorker(HWND hwndParent, DWORD dwExStyle, DWORD dwStyle)
+{
+    m_hWnd = SHCreateWorkerWindowW(WindowProc, hwndParent, dwExStyle, dwStyle,
+                                   NULL, (LONG_PTR)this);
+    return m_hWnd != NULL;
+}
+
+struct CChangeNotifyImpl
+{
+    typedef CChangeNotify::ITEM ITEM;
+    CSimpleArray<ITEM> m_items;
+
+    BOOL AddItem(UINT nRegID, DWORD dwUserPID, HANDLE hShare)
+    {
+        for (INT i = 0; i < m_items.GetSize(); ++i)
+        {
+            if (m_items[i].nRegID == INVALID_REG_ID)
+            {
+                m_items[i].nRegID = nRegID;
+                m_items[i].dwUserPID = dwUserPID;
+                m_items[i].hShare = hShare;
+                return TRUE;
+            }
+        }
+
+        ITEM item = { nRegID, dwUserPID, hShare };
+        m_items.Add(item);
+        return TRUE;
+    }
+
+    BOOL RemoveItem(UINT nRegID, DWORD dwOwnerPID)
+    {
+        BOOL bFound = FALSE;
+        for (INT i = 0; i < m_items.GetSize(); ++i)
+        {
+            if (m_items[i].nRegID == nRegID)
+            {
+                bFound = TRUE;
+                SHFreeShared(m_items[i].hShare, dwOwnerPID);
+                m_items[i].nRegID = INVALID_REG_ID;
+                m_items[i].dwUserPID = 0;
+                m_items[i].hShare = NULL;
+            }
+        }
+        return bFound;
+    }
+
+    void RemoveItemsByProcess(DWORD dwOwnerPID, DWORD dwUserPID)
+    {
+        for (INT i = 0; i < m_items.GetSize(); ++i)
+        {
+            if (m_items[i].dwUserPID == dwUserPID)
+            {
+                SHFreeShared(m_items[i].hShare, dwOwnerPID);
+                m_items[i].nRegID = INVALID_REG_ID;
+                m_items[i].dwUserPID = 0;
+                m_items[i].hShare = NULL;
+            }
+        }
+    }
+};
+
+CChangeNotify::CChangeNotify()
+    : m_nNextRegID(INVALID_REG_ID)
+    , m_pimpl(new CChangeNotifyImpl)
+{
+}
+
+CChangeNotify::~CChangeNotify()
+{
+    delete m_pimpl;
+}
+
+void CChangeNotify::clear()
+{
+    m_hWnd = NULL;
```

Comment:
Who destroys the window, if there is one?

Reply:
`CChangeNotify::clear()` is deleted.